connect/proxy/proxy.go

```diff
@@ -126,6 +137,30 @@ func (p *Proxy) Serve() error {
 	}
 }
 
+// removeOldListeners removes listeners from `active` that aren't referenced in newUpstreams. The
+// removed listeners are closed and any errors are logged. This should be called BEFORE adding new
+// listeners from a new configuration so that if a particular port is switched from one target
+// service to another service it will correctly be re-tasked.
+func (p *Proxy) removeOldListeners(active map[string]*Listener, newUpstreams []UpstreamConfig) {
+	oldUpstreamsToRemove := map[string]bool{}
+	// Initially assume that we should remove all of the active upstreams.
+	for active := range active {
+		oldUpstreamsToRemove[active] = true
+	}
+	// Mark all the new configurations so that we don't remove one that hasn't changed.
+	for _, uc := range newUpstreams {
+		delete(oldUpstreamsToRemove, uc.String())
+	}
+	// Actually remove the upstreams that aren't referenced in the new config.
+	for upstream := range oldUpstreamsToRemove {
+		p.logger.Info("Removing now-unused listener", "listener", upstream)
+		if err := active[upstream].Close(); err != nil {
+			p.logger.Error("failed closing old upstream", "listener", upstream, "error", "err")
+		}
+		delete(active, upstream)
```

Comment:
If you can't close it, should it stay in active?

Reply:
hrm.... no, I think not, but I'll add a comment.

Reasoning: If we get an error closing it, then removing it from active will mean:
* we won't try to re-use it, which seems right if we've tried to close it
* instead, we'll potentially add a new listener later that listens to the same port. This might fail (if the port is still bound) which is better than just using the wrong Listener.

It's a moot point at the moment because `l.Close()` will never return a non-nil error. I very nearly just omitted the error-handling, but couldn't bring myself to do that.